File: commonutils/src/main/java/net/maxsmr/commonutils/logger/holder/BaseLoggerHolder.java
package net.maxsmr.commonutils.logger.holder;

import android.text.TextUtils;

import net.maxsmr.commonutils.logger.BaseLogger;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

import static net.maxsmr.commonutils.data.text.TextUtilsKt.isEmpty;

public abstract class BaseLoggerHolder {

    private static BaseLoggerHolder sInstance;

    public static BaseLoggerHolder getInstance() {
        synchronized (BaseLoggerHolder.class) {
            if (sInstance == null) {
                throw new IllegalStateException(BaseLoggerHolder.class.getSimpleName() + " is not initialized");
            }
            return sInstance;
        }
    }

    public static void initInstance(@NotNull ILoggerHolderProvider<?> provider) {
        synchronized (BaseLoggerHolder.class) {
            if (sInstance == null) {
                sInstance = provider.provideHolder();
            }
        }
    }

    public static void releaseInstance() {
        synchronized (BaseLoggerHolder.class) {
            sInstance = null;
        }
    }

    private final Map<String, BaseLogger> loggersMap = new LinkedHashMap<>();

    private final boolean isNullInstancesAllowed;

    protected BaseLoggerHolder(boolean isNullInstancesAllowed) {
        this.isNullInstancesAllowed = isNullInstancesAllowed;
    }

    public Map<String, BaseLogger> getLoggersMap() {
        synchronized (loggersMap) {
            return Collections.unmodifiableMap(loggersMap);
        }
    }

    public int getLoggersCount() {
        return getLoggersMap().size();
    }

    public boolean isNullInstancesAllowed() {
        return isNullInstancesAllowed;
    }

    /**
     * @param clazz object class to get/create logger for
     * */
    @SuppressWarnings("unchecked")
    public <L extends BaseLogger> L getLogger(Class<?> clazz) {
        return getLogger(clazz.getName());
    }

    /**
     * @param className object class to get/create logger for
     * */
    @SuppressWarnings("unchecked")
    public <L extends BaseLogger> L getLogger(String className) {
        if (isEmpty(className)) {
            throw new IllegalArgumentException("className is empty");
        }
        synchronized (loggersMap) {
            BaseLogger logger = loggersMap.get(className);
            if (logger == null) {
                boolean addToMap = true;
                logger = createLogger(className);
                if (logger == null) {
                    if (!isNullInstancesAllowed) {
                        throw new RuntimeException("Logger was not created for className: " + className);
                    }
                    logger = new BaseLogger.Stub();
                    addToMap = false;
                }
                if (addToMap) {
                    loggersMap.put(className, logger);
                }
            }
            return (L) logger;
        }
    }

    @Nullable
    protected abstract BaseLogger createLogger(@NotNull String className);

    public static void logException(@NotNull BaseLogger logger, @NotNull Exception e) {
        logException(logger, e, null);
    }

    public static void logException(@NotNull BaseLogger logger, @NotNull Exception e, @Nullable String description) {
        logger.e(formatException(e, description), e);
    }

    public static String formatException(@NotNull Exception e) {
        return formatException(e, null);
    }

    public static String formatException(@NotNull Exception e, @Nullable String description) {
        return "A(n) " + e.getClass().getSimpleName() + " occurred"
                + (TextUtils.isEmpty(description)? "" : " during " + description) + ": " + e.getMessage();
    }

    public static void throwRuntimeException(@NotNull Exception e) throws RuntimeException {
        throwRuntimeException(e, null);
    }

    public static void throwRuntimeException(@NotNull Exception e, @Nullable String description) throws RuntimeException {
        throw new RuntimeException(formatException(e, description), e);
    }
}

File: commonutils-jre/src/main/kotlin/net/maxsmr/commonutils/logger/holder/BaseLoggerHolder.kt
package net.maxsmr.commonutils.logger.holder

import net.maxsmr.commonutils.logger.BaseLogger
import net.maxsmr.commonutils.text.EMPTY_STRING
import net.maxsmr.commonutils.text.isEmpty
import java.util.*
import kotlin.collections.HashMap

abstract class BaseLoggerHolder protected constructor() {

    private val loggersMap =  Collections.synchronizedMap(HashMap<String, BaseLogger>())

    val loggersCount: Int get() = loggersMap.size

    /**
     * @param clazz object class to get/create logger for
     */
    fun <L : BaseLogger?> getLogger(clazz: Class<*>): L =
            getLogger(clazz.name)

    /**
     * @param className object class to get/create logger for
     */
    @Suppress("UNCHECKED_CAST")
    fun <L : BaseLogger> getLogger(className: String): L {
        require(className.isNotEmpty()) { "className is empty" }
            var logger = loggersMap[className]
            if (logger == null) {
                var addToMap = true
                logger = createLogger(className)
                if (logger == null) {
                    logger = BaseLogger.Stub(EMPTY_STRING)
                    addToMap = false
                }
                if (addToMap) {
                    loggersMap[className] = logger
                }
            }
            return logger as L
    }

    protected abstract fun createLogger(className: String): BaseLogger?

    companion object {

        @JvmStatic
        val instance: BaseLoggerHolder
            get() {
                synchronized(BaseLoggerHolder::class.java) {
                    with(sInstance) {
                        checkNotNull(this) { "LoggerHolder is not initialized" }
                        return this
                    }
                }
            }

        private var sInstance: BaseLoggerHolder? = null

        @JvmStatic
        fun initInstance(provider: () -> BaseLoggerHolder) {
            synchronized(BaseLoggerHolder::class.java) {
                if (sInstance == null) {
                    sInstance = provider()
                }
            }
        }

        @JvmStatic
        fun releaseInstance() {
            synchronized(BaseLoggerHolder::class.java) { sInstance = null }
        }

        @JvmStatic
        @JvmOverloads
        fun logException(logger: BaseLogger, e: Throwable, description: String? = null) {
            logger.e(formatException(e, description), e)
        }

        @JvmStatic
        @JvmOverloads
        fun formatException(e: Throwable, description: String? = null): String {
            return ("A(n) ${e.javaClass.simpleName} occurred"
                    + (if (isEmpty(description)) EMPTY_STRING else " by $description") + ": ${e.message}")
        }

        @JvmStatic
        @JvmOverloads
        @Throws(RuntimeException::class)
        fun throwRuntimeException(e: Throwable, description: String? = null) {
            throw RuntimeException(formatException(e, description), e)
        }
    }
}
